slashing/program/src/instruction.rs

```diff
@@ -0,0 +1,146 @@
+//! Program instructions
+
+use {
+    crate::id,
+    solana_program::{
+        clock::Slot,
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+    },
+    std::mem::size_of,
+};
+
+/// Instructions supported by the program
+#[derive(Clone, Debug, PartialEq)]
+pub enum SlashingInstruction {
+    /// Submit a slashable violation proof for `node_pubkey`, which indicates
+    /// that they submitted a duplicate block to the network
+    ///
+    ///
+    /// Accounts expected by this instruction:
+    /// 0. `[]` Proof account, must be previously initialized with the proof
+    ///    data.
+    ///
+    /// We expect the proof account to be properly sized as to hold a duplicate
+    /// block proof. See [ProofType] for sizing requirements.
+    ///
+    /// Deserializing the proof account from `offset` should result in a
+    /// [DuplicateBlockProofData]
+    DuplicateBlockProof {
+        /// Offset into the proof account to begin reading, expressed as `u64`
+        offset: u64,
+        /// Slot for which the violation occured
+        slot: Slot,
+        /// Identity pubkey of the Node that signed the duplicate block
+        node_pubkey: Pubkey,
+    },
+}
+
+impl SlashingInstruction {
```

Comment:
very cool, I added that here d19b90a100d30dbf6129b8a1c73462da9def8fd3

ended up copying over some of the helpers because I figured adding a dependency on the token program is not great?